lib/slax_web/websocket_listener.ex

```diff
@@ -0,0 +1,77 @@
+defmodule SlaxWeb.WebsocketListener do
+  use GenServer
+
+  alias Slax.Http
+  alias SlaxWeb.Issue
+  alias SlaxWeb.Poker
+
+  defp config() do
+    Application.get_env(:slax, Slax.Slack)
+  end
+
+  defp app_token() do
+    config()[:app_token]
+  end
+
+  def init(_) do
+    {:ok, %{body: %{"url" => "wss://wss-primary.slack.com" <> path}}} =
+      Http.post(
+        "https://slack.com/api/apps.connections.open",
+        "",
+        "Content-Type": "application/x-www-form-urlencoded",
+        Authorization: "Bearer #{app_token()}"
+      )
+
+    {:ok, pid} =
+      :gun.open(:binary.bin_to_list("wss-primary.slack.com"), 443, %{
+        transport: :tls,
+        protocols: [:http],
+        tls_opts: [verify: :verify_none, log_level: :debug]
+      })
+
+    {:ok, :http} = :gun.await_up(pid, 10_000)
+    stream = :gun.ws_upgrade(pid, path)
+    {:upgrade, [<<"websocket">>], _headers} = :gun.await(pid, stream, 10_000)
+
+    {:ok, nil}
+  end
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, nil)
+  end
+
+  def handle_info({_, pid, stream_ref, {:text, event}}, socket) do
+    {:ok, decoded_event} = Jason.decode(event)
+    handle_message(pid, stream_ref, decoded_event)
+    {:noreply, socket}
+  end
+
+  # Handle disconnects
+  defp handle_message(pid, _, %{"type" => "disconnect"}) do
+    :gun.close(pid)
+    init(nil)
+  end
+
+  # Websocket init message
+  defp handle_message(_, _, %{"type" => "hello"}), do: nil
+
+  defp handle_message(pid, stream_ref, %{
+         "type" => "events_api",
+         "envelope_id" => envelope_id,
+         "payload" => %{"event" => event}
+       }) do
+    Issue.handle_event(event)
+
+    response = Jason.encode!(%{envelope_id: envelope_id})
```

Comment:
is there a particular reason why sometimes using ! and other times not? My inclination is to try to handle them all consistently and predictably

Reply:
No that was just my own oversight, changing them all to use !